# TASK 1 : Install Dependencies & Run a SparkSession


In [ ]:
#install pyspark
! pip install pyspark

In [ ]:
#create a sparksessio
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName('spark').getOrCreate()

#getOrCreate checks if an existing session with the same name is present, if not it creates a new session

# TASK 2 : Clone & Explore dataset

In [ ]:
#clone the dataset
! git clone https://github.com/education454/admission_dataset

Cloning into 'admission_dataset'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (3/3), 5.60 KiB | 5.60 MiB/s, done.


In [ ]:
#check the presence of dataset
! ls admission_dataset

Admission_Predict_Ver1.1.csv


In [ ]:
#create a spark dataframe
df = spark.read.csv('/content/admission_dataset/Admission_Predict_Ver1.1.csv' , header = True , inferSchema= True)
#inferschema auto checks the datatypes

In [ ]:
#display dataframe
df.show()

+---------+---------+-----------+-----------------+---+---+----+--------+---------------+
|Serial No|GRE Score|TOEFL Score|University Rating|SOP|LOR|CGPA|Research|Chance of Admit|
+---------+---------+-----------+-----------------+---+---+----+--------+---------------+
|        1|      337|        118|                4|4.5|4.5|9.65|       1|           0.92|
|        2|      324|        107|                4|4.0|4.5|8.87|       1|           0.76|
|        3|      316|        104|                3|3.0|3.5| 8.0|       1|           0.72|
|        4|      322|        110|                3|3.5|2.5|8.67|       1|            0.8|
|        5|      314|        103|                2|2.0|3.0|8.21|       0|           0.65|
|        6|      330|        115|                5|4.5|3.0|9.34|       1|            0.9|
|        7|      321|        109|                3|3.0|4.0| 8.2|       1|           0.75|
|        8|      308|        101|                2|3.0|4.0| 7.9|       0|           0.68|
|        9

In [ ]:
#get the no.of  columns
print(df.count(),len(df.columns))

500 9


In [ ]:
#print schema
df.printSchema()

root
 |-- Serial No: integer (nullable = true)
 |-- GRE Score: integer (nullable = true)
 |-- TOEFL Score: integer (nullable = true)
 |-- University Rating: integer (nullable = true)
 |-- SOP: double (nullable = true)
 |-- LOR: double (nullable = true)
 |-- CGPA: double (nullable = true)
 |-- Research: integer (nullable = true)
 |-- Chance of Admit: double (nullable = true)



In [ ]:
#get the summary statistics
df.describe().show()


+-------+-----------------+------------------+-----------------+-----------------+------------------+------------------+------------------+------------------+-------------------+
|summary|        Serial No|         GRE Score|      TOEFL Score|University Rating|               SOP|               LOR|              CGPA|          Research|    Chance of Admit|
+-------+-----------------+------------------+-----------------+-----------------+------------------+------------------+------------------+------------------+-------------------+
|  count|              500|               500|              500|              500|               500|               500|               500|               500|                500|
|   mean|            250.5|           316.472|          107.192|            3.114|             3.374|             3.484| 8.576440000000003|              0.56| 0.7217399999999996|
| stddev|144.4818327679989|11.295148372354712|6.081867659564538|1.143511800759815|0.9910036207566072|0.92

# TASK 3 : Data Cleaning

In [ ]:
#drop the unnecessary column
df = df.drop('Serial No')


In [ ]:
#display the dataframe
df.show()

+---------+-----------+-----------------+---+---+----+--------+---------------+
|GRE Score|TOEFL Score|University Rating|SOP|LOR|CGPA|Research|Chance of Admit|
+---------+-----------+-----------------+---+---+----+--------+---------------+
|      337|        118|                4|4.5|4.5|9.65|       1|           0.92|
|      324|        107|                4|4.0|4.5|8.87|       1|           0.76|
|      316|        104|                3|3.0|3.5| 8.0|       1|           0.72|
|      322|        110|                3|3.5|2.5|8.67|       1|            0.8|
|      314|        103|                2|2.0|3.0|8.21|       0|           0.65|
|      330|        115|                5|4.5|3.0|9.34|       1|            0.9|
|      321|        109|                3|3.0|4.0| 8.2|       1|           0.75|
|      308|        101|                2|3.0|4.0| 7.9|       0|           0.68|
|      302|        102|                1|2.0|1.5| 8.0|       0|            0.5|
|      323|        108|                3

In [ ]:
#check for null values
for i in df.columns:
  print(i+ ":" , df[df[i].isNull()].count())

GRE Score: 0
TOEFL Score: 0
University Rating: 0
SOP: 0
LOR: 0
CGPA: 0
Research: 0
Chance of Admit: 0


# TASK 4 : Correlation Analysis & Feature Selection

In [ ]:
# correlation analysis
for i in df.columns:
  print('Correlation to chance of Admit col for {} is {}'.format(i,df.stat.corr('Chance of Admit' , i)))


Correlation to chance of Admit col for GRE Score is 0.8103506354632598
Correlation to chance of Admit col for TOEFL Score is 0.7922276143050823
Correlation to chance of Admit col for University Rating is 0.6901323687886892
Correlation to chance of Admit col for SOP is 0.6841365241316723
Correlation to chance of Admit col for LOR is 0.6453645135280112
Correlation to chance of Admit col for CGPA is 0.882412574904574
Correlation to chance of Admit col for Research is 0.5458710294711379
Correlation to chance of Admit col for Chance of Admit is 1.0


Pick the column with the highest correlation score which in my case is :
* GRE => 0.8103506354632598
* CGPA =>  0.882412574904574
* TOEFL => 0.7922276143050823

In [ ]:
# feature selection
from pyspark.ml.feature import VectorAssembler
# transformation -> to merge multiple columns into a vector column
assembler = VectorAssembler(inputCols=['GRE Score' , 'TOEFL Score' , 'CGPA' ] , outputCol='features')

In [ ]:
#display dataframe
output_data = assembler.transform(df)
output_data.show()

+---------+-----------+-----------------+---+---+----+--------+---------------+------------------+
|GRE Score|TOEFL Score|University Rating|SOP|LOR|CGPA|Research|Chance of Admit|          features|
+---------+-----------+-----------------+---+---+----+--------+---------------+------------------+
|      337|        118|                4|4.5|4.5|9.65|       1|           0.92|[337.0,118.0,9.65]|
|      324|        107|                4|4.0|4.5|8.87|       1|           0.76|[324.0,107.0,8.87]|
|      316|        104|                3|3.0|3.5| 8.0|       1|           0.72| [316.0,104.0,8.0]|
|      322|        110|                3|3.5|2.5|8.67|       1|            0.8|[322.0,110.0,8.67]|
|      314|        103|                2|2.0|3.0|8.21|       0|           0.65|[314.0,103.0,8.21]|
|      330|        115|                5|4.5|3.0|9.34|       1|            0.9|[330.0,115.0,9.34]|
|      321|        109|                3|3.0|4.0| 8.2|       1|           0.75| [321.0,109.0,8.2]|
|      308

# TASK 5 : Build the Linear Regression Model

In [ ]:
#import Linearregression and create final data
from pyspark.ml.regression import LinearRegression
final_data = output_data.select('features' , 'Chance of Admit')

In [ ]:
#print schema of final data
final_data.printSchema()

root
 |-- features: vector (nullable = true)
 |-- Chance of Admit: double (nullable = true)



In [ ]:
#split the dataset into training and testing set
train , test = final_data.randomSplit([0.7,0.3])

In [ ]:
#build & train the model
models = LinearRegression(featuresCol='features' , labelCol='Chance of Admit')
model = models.fit(train)


In [ ]:
#get coefficients & intercept
print('coefficents:' , model.coefficients)
print('intercept:' , model.intercept)

coefficents: [0.0020182916153442484,0.0025620949324366006,0.15462131150475644]
intercept: -1.5181070071640972


In [ ]:
#get summary of the model
summary = model.summary

In [ ]:
#print the rmse & r2 score
print('RMSE' , summary.rootMeanSquaredError) # error within the predicted value and the observed value (how well it performed) lesser the better
print('r2' , summary.r2) #how close it is to the fitted line -> higher the better

RMSE 0.06200610035561955
r2 0.8030872179674469


# TASK 6 : Evaluate & Save the Model

In [ ]:
#transform on the test data
predictions = model.transform(test)

In [ ]:
#display the predictions
predictions.show(20)

+------------------+---------------+-------------------+
|          features|Chance of Admit|         prediction|
+------------------+---------------+-------------------+
|[290.0,100.0,7.56]|           0.47|0.49234416950535365|
| [294.0,95.0,7.64]|           0.49|0.49997656622492825|
| [296.0,95.0,7.54]|           0.44| 0.4885510183051409|
| [297.0,96.0,7.89]|           0.43| 0.5472488638795865|
| [297.0,98.0,7.67]|           0.59| 0.5183563652134133|
| [298.0,92.0,7.88]|           0.51| 0.5374725626501367|
|  [298.0,98.0,7.5]|           0.44| 0.4940890338729489|
| [298.0,98.0,8.03]|           0.34| 0.5760383289704698|
| [298.0,99.0,7.46]|           0.53|0.49046627634519546|
|[298.0,100.0,7.95]|           0.58| 0.5687928139149625|
| [299.0,94.0,7.34]|           0.42|0.46111953591778576|
|[299.0,100.0,7.89]|           0.59| 0.5615338268400214|
| [299.0,106.0,8.4]|           0.64| 0.6557632653020671|
| [300.0,95.0,8.22]|           0.62| 0.6017666765897525|
|[300.0,100.0,8.26]|           

In [ ]:
#evaluate the model

from pyspark.ml.evaluation import RegressionEvaluator
evaluator = RegressionEvaluator(predictionCol ='prediction' , labelCol ='Chance of Admit' , metricName = 'r2')
print('r2 on the test data' , evaluator.evaluate(predictions))


r2 on the test data 0.8056535864834586


In [ ]:
#save the model
model.save("model")

In [ ]:
#load the model
from pyspark.ml.regression import LinearRegressionModel
model = LinearRegressionModel.load("model")
